In [14]:
import sqlalchemy as alch
import os
import dotenv
dotenv.load_dotenv()
import pandas as pd
import sys
sys.path.append('../src')
import apilast as als
import sqltools as sqt
import requests
import datetime

In [3]:
passw = os.getenv("mysql")
dbName = "visuallast"
connectionData = f"mysql+pymysql://root:{passw}@localhost/{dbName}"
engine = alch.create_engine(connectionData)

In [6]:
keylast = os.getenv("keylast")
l_user = 'sinatxester'

## base de datos para insertar scrobblings de los usuarios elegidos
- vemos aquí las tablas creadas

```mysql
#country, playcount,realname, registered['unixtime'],url,gender, name, type
create table usuarios (id_us int primary key auto_increment,
					nombre varchar(50) unique, #name
                    country varchar(50), #podemos modificarlo luego para que sea un fk.
                    playcount int, # habrá que actualizarlo... también podemos sacarlo de los scrobblings
                    image varchar(100), #imagemedium 
                    realname varchar(50),
                    fechahora datetime, # verndría de unixtime con una función aplicada al insertar datos
                    url varchar(150),
                    gender varchar(5), # no sé si es como con la edad que no aparece nunca.
                    typeuser varchar(15)
                    );
#scrobblings vienen de los recentracks
create table scrobblings (id_scro int primary key auto_increment,
						artist_name varchar(100),
                        artist_mbid varchar(50),
                        album_name varchar(255),
                        album_mbid varchar(50),
                        title text,
                        title_mbid varchar(50),
                        url varchar(255),
                        fecharhoa datetime, # viene de unixtime al que hay que aplicar una funcióncita
                        id_us int,
                        CONSTRAINT fk_scro_user FOREIGN KEY (id_us)  REFERENCES usuarios(id_us)  
						);
                        
# friends de friends
create table amiguismo (id_am int primary key auto_increment,
						nombre varchar(50), #usuario principal
                        amigo varchar(50), #nombre del amigo
                        constraint fk_amigo_usuario foreign key (nombre) references usuarios(nombre)
                        );


                    

```

In [7]:
url_user = f'http://ws.audioscrobbler.com/2.0/?method=user.getinfo&user={l_user}&api_key={keylast}&format=json'

In [11]:
req_user = requests.get(url_user).json()['user']
req_user

{'country': 'Spain',
 'age': '0',
 'playcount': '77557',
 'subscriber': '0',
 'realname': 'Ester Sinatxe',
 'playlists': '0',
 'bootstrap': '0',
 'image': [{'size': 'small',
   '#text': 'https://lastfm.freetls.fastly.net/i/u/34s/daa41927775a61b3e4762c4a8eda80f3.png'},
  {'size': 'medium',
   '#text': 'https://lastfm.freetls.fastly.net/i/u/64s/daa41927775a61b3e4762c4a8eda80f3.png'},
  {'size': 'large',
   '#text': 'https://lastfm.freetls.fastly.net/i/u/174s/daa41927775a61b3e4762c4a8eda80f3.png'},
  {'size': 'extralarge',
   '#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/daa41927775a61b3e4762c4a8eda80f3.png'}],
 'registered': {'unixtime': '1420141064', '#text': 1420141064},
 'url': 'https://www.last.fm/user/sinatxester',
 'gender': 'n',
 'name': 'sinatxester',
 'type': 'user'}

In [12]:
def car_esp(palabrita):
    '''
    recibe un elemento, si no es string lo convierte
    comprueba si en la cadena de texto existen una serie de caracteres
    si estos caracteres están en la cadena los reemplaza
    devuelve la palabrita
    '''
    palabrita = str(palabrita)
    dicc = {'\\':'\\\\',"'":"\\'",'"':'\\"',"%":"%%"}
    if "'" in palabrita or '"' in palabrita  or '\\' in palabrita or '%' in palabrita:
        for k,v in dicc.items():
            palabrita = palabrita.replace(k,v)
        return palabrita
    else:
        return palabrita

In [13]:
def utslocal (uts):
    '''
    recibe una cadena de texto o interger que es uts(unix time stamp, valor en segundos de una fecha)
        unix timestamp count = This count starts at the Unix Epoch on January 1st, 1970 at UTC
    devuelve fecha y hora local. 
    '''
    if type(uts) == int:
        return datetime.datetime.fromtimestamp(uts).strftime('%Y-%m-%d %H:%M:%S')
    else:
        return datetime.datetime.fromtimestamp(int(uts)).strftime('%Y-%m-%d %H:%M:%S')

In [ ]:
def check_user (l_user):
    

In [22]:
# definimos la función para insertar los datos del usuario
def insert_user(l_user):
    # recibe nombre de usuario
    # llama a la api de lastfm
    # inserta los datos del usuario en la tabla usuarios de mysql
    print(f'llamando a la api para extraer los datos del usuario {l_user}')
    keylast = os.getenv("keylast")
    url = f'http://ws.audioscrobbler.com/2.0/?method=user.getinfo&user={l_user}&api_key={keylast}&format=json'
    req = requests.get(url).json()['user']
    nombre = car_esp(req['name'])
    country = req['country']
    image = car_esp(req['image'][2]['#text'])
    realname = car_esp(req['realname'])
    fechahora = utslocal(req['registered']['unixtime'])
    urluser = car_esp(req['url'])
    gender = req['gender']
    typeuser = req['type']
    print(nombre,country,image,realname,fechahora,urluser,gender,typeuser)
    print('insertando datos en mysql')
    engine.execute(f'''
            INSERT INTO usuarios (nombre, country, image, realname, fechahora, url, gender, typeuser)
            VALUES('{nombre}','{country}','{image}','{realname}','{fechahora}','{urluser}','{gender}','{typeuser}')
            ''')
    print('datos insertados en mysql')

In [23]:
insert_user('sinatxester')

llamando a la api para extraer los datos del usuario sinatxester
sinatxester Spain https://lastfm.freetls.fastly.net/i/u/174s/daa41927775a61b3e4762c4a8eda80f3.png Ester Sinatxe 2015-01-01 20:37:44 https://www.last.fm/user/sinatxester n user
insertando datos en mysql


IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry 'sinatxester' for key 'usuarios.nombre'")
[SQL: 
            INSERT INTO usuarios (nombre, country, image, realname, fechahora, url, gender, typeuser)
            VALUES('sinatxester','Spain','https://lastfm.freetls.fastly.net/i/u/174s/daa41927775a61b3e4762c4a8eda80f3.png','Ester Sinatxe','2015-01-01 20:37:44','https://www.last.fm/user/sinatxester','n','user')
            ]
(Background on this error at: https://sqlalche.me/e/14/gkpj)

- la función de insertuser debería checkear primero si el usuario existe, poerque al ser unique no va a dejar insertarlo

# IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry 'sinatxester' for key 'usuarios.nombre'") 